{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sqlite3\n",
    "import pandas as pd\n",
    "import os\n",
    "import sys\n",
    "from Bio.SubsMat import MatrixInfo as matlist\n",
    "from Bio import pairwise2\n",
    "from Bio.pairwise2 import format_alignment\n",
    "import math\n",
    "import numpy as np\n",
    "def fasta2List(pathFasta):\n",
    "    f = open(pathFasta, \"r\")\n",
    "    title = []\n",
    "    seq = []\n",
    "    seq_temp = []\n",
    "    for line in f:\n",
    "        if line[0] == \">\":\n",
    "            seq.append(''.join(seq_temp).replace(\"\\n\", \"\"))\n",
    "            title.append(line.replace(\"\\n\", \"\"))\n",
    "            seq_temp = []\n",
    "        else:\n",
    "            seq_temp.append(line)\n",
    "    seq.append(''.join(seq_temp).replace(\"\\n\", \"\"))\n",
    "    seq.pop(0)\n",
    "    dictionary = dict(zip(title, seq))\n",
    "    return dictionary\n",
    "\n",
    "conn = sqlite3.connect('../../mismatch_db.db')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "mismatch = pd.read_sql_query(\"SELECT * FROM mismatch\", conn)\n",
    "mismatch = mismatch.astype({\"exon_start_prim\": \"Int64\", \"exon_stop_prim\":\"Int64\", \"exon_start_hum\": \"Int64\", \"exon_stop_hum\":\"Int64\"})\n",
    "\n",
    "prim_exon_introns = pd.read_sql_query(\"\"\"\n",
    "SELECT mismatch_ID, mismatch.prot_prim, exon_intron_map.'type', exon_intron_map.number_elem, exon_intron_map.seq\n",
    "FROM mismatch\n",
    "JOIN protein ON mismatch.prot_prim = protein.prot_ID\n",
    "JOIN transcript ON protein.transcript_ID = transcript.transcript_ID\n",
    "JOIN exon_intron_map ON transcript.transcript_ID = exon_intron_map.transcript_ID\n",
    "\"\"\", conn)\n",
    "\n",
    "mismatch_prot_transcript = pd.read_sql_query(\n",
    "\"\"\"SELECT * \n",
    "FROM mismatch\n",
    "JOIN protein ON mismatch.prot_prim = protein.prot_ID\n",
    "JOIN transcript ON protein.transcript_ID = transcript.transcript_ID\"\"\", conn)\n",
    "\n",
    "mismatch_to_correct = pd.read_sql_query(\n",
    "\"\"\"SELECT mismatch.*\n",
    "FROM mismatch_flag \n",
    "JOIN tblastn_match ON mismatch_flag.mismatch_ID = tblastn_match.mismatch_ID \n",
    "JOIN mismatch on mismatch_flag.mismatch_ID = mismatch.mismatch_ID \n",
    "WHERE (one_hum_multiple_prim = 1\n",
    "OR non_canonical_hum_spl = 1\n",
    "OR N_in_genomic = 1\n",
    "OR small_introns = 1)\n",
    "AND conserved = 0\n",
    "AND repeats_prot = 0\n",
    "AND alignement_error = 0\n",
    "AND human_isoform_exist = 0\n",
    "\"\"\", conn)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mismatch_ID</th>\n",
       "      <th>prot_prim</th>\n",
       "      <th>type</th>\n",
       "      <th>number_elem</th>\n",
       "      <th>seq</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>A0A0D9RC82</td>\n",
       "      <td>Exon</td>\n",
       "      <td>1</td>\n",
       "      <td>CACCGCTGGCCAGACGAGGGATGTTATTTTGGGCAGTGCATCCGGA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0</td>\n",
       "      <td>A0A0D9RC82</td>\n",
       "      <td>Exon</td>\n",
       "      <td>2</td>\n",
       "      <td>CATCTACAAATCTGAAAGACAAAACATGGTTCAAACATCCAGGCAC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0</td>\n",
       "      <td>A0A0D9RC82</td>\n",
       "      <td>Exon</td>\n",
       "      <td>3</td>\n",
       "      <td>GTCCACCCGTGGCTCCAAAATGGTCTCCTGGTCCGTGACAGCAAAG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0</td>\n",
       "      <td>A0A0D9RC82</td>\n",
       "      <td>Exon</td>\n",
       "      <td>4</td>\n",
       "      <td>GTGTTCGGCCTTGGTTCCGTGGCCCATATGGTTCTAAATAAAAAAT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0</td>\n",
       "      <td>A0A0D9RC82</td>\n",
       "      <td>Exon</td>\n",
       "      <td>5</td>\n",
       "      <td>GGGCCCACATGAACGCAGCTGTGACCTTCGCTAACTGTGCACTGGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>383254</th>\n",
       "      <td>11014</td>\n",
       "      <td>H0WR01</td>\n",
       "      <td>Intron</td>\n",
       "      <td>17</td>\n",
       "      <td>GTAAGTTTATATATTGACTAGGGCTTCATCTTTTATTTTGGAATCA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>383255</th>\n",
       "      <td>11014</td>\n",
       "      <td>H0WR01</td>\n",
       "      <td>Intron</td>\n",
       "      <td>18</td>\n",
       "      <td>GTGAGTGATGTGAGTTTACTTCCTTTTCTGAAGATGATGATGAACT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>383256</th>\n",
       "      <td>11014</td>\n",
       "      <td>H0WR01</td>\n",
       "      <td>Intron</td>\n",
       "      <td>19</td>\n",
       "      <td>GTAAAAAGCCAATACTATAAAGCTGTGTTCCCTTCCTGTCATTTTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>383257</th>\n",
       "      <td>11014</td>\n",
       "      <td>H0WR01</td>\n",
       "      <td>Intron</td>\n",
       "      <td>20</td>\n",
       "      <td>GTAAATTTGTCATTTGAGGCATTTTCCAATACAAATATTGTTTATA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>383258</th>\n",
       "      <td>11014</td>\n",
       "      <td>H0WR01</td>\n",
       "      <td>Intron</td>\n",
       "      <td>21</td>\n",
       "      <td>GTAAGTATTCTTAACTAAAGTAACATTTGAGTGGTTTTTAGGTGCT...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>383259 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        mismatch_ID   prot_prim    type  number_elem  \\\n",
       "0                 0  A0A0D9RC82    Exon            1   \n",
       "1                 0  A0A0D9RC82    Exon            2   \n",
       "2                 0  A0A0D9RC82    Exon            3   \n",
       "3                 0  A0A0D9RC82    Exon            4   \n",
       "4                 0  A0A0D9RC82    Exon            5   \n",
       "...             ...         ...     ...          ...   \n",
       "383254        11014      H0WR01  Intron           17   \n",
       "383255        11014      H0WR01  Intron           18   \n",
       "383256        11014      H0WR01  Intron           19   \n",
       "383257        11014      H0WR01  Intron           20   \n",
       "383258        11014      H0WR01  Intron           21   \n",
       "\n",
       "                                                      seq  \n",
       "0       CACCGCTGGCCAGACGAGGGATGTTATTTTGGGCAGTGCATCCGGA...  \n",
       "1       CATCTACAAATCTGAAAGACAAAACATGGTTCAAACATCCAGGCAC...  \n",
       "2       GTCCACCCGTGGCTCCAAAATGGTCTCCTGGTCCGTGACAGCAAAG...  \n",
       "3       GTGTTCGGCCTTGGTTCCGTGGCCCATATGGTTCTAAATAAAAAAT...  \n",
       "4       GGGCCCACATGAACGCAGCTGTGACCTTCGCTAACTGTGCACTGGG...  \n",
       "...                                                   ...  \n",
       "383254  GTAAGTTTATATATTGACTAGGGCTTCATCTTTTATTTTGGAATCA...  \n",
       "383255  GTGAGTGATGTGAGTTTACTTCCTTTTCTGAAGATGATGATGAACT...  \n",
       "383256  GTAAAAAGCCAATACTATAAAGCTGTGTTCCCTTCCTGTCATTTTG...  \n",
       "383257  GTAAATTTGTCATTTGAGGCATTTTCCAATACAAATATTGTTTATA...  \n",
       "383258  GTAAGTATTCTTAACTAAAGTAACATTTGAGTGGTTTTTAGGTGCT...  \n",
       "\n",
       "[383259 rows x 5 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "prim_exon_introns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Blast pos x3 - 3\n",
    "# 1/ Enlever le partie codant pour le mismatch du transcript\n",
    "\n",
    "# 2/ Ajouter un nouvel exon codant pour la bonne séquence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 182,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/meyer/anaconda3/envs/stage-env/lib/python3.7/site-packages/ipykernel_launcher.py:33: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "/home/meyer/anaconda3/envs/stage-env/lib/python3.7/site-packages/ipykernel_launcher.py:34: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DELETE THE END OF EXON\n",
      "DELETE THE BEGIN OF EXON\n",
      "***\n",
      "DELETE INSIDE THE EXON\n",
      "***\n",
      "TO DELETE TOTALLY\n",
      "TO DELETE TOTALLY\n",
      "TO DELETE TOTALLY\n",
      "TO DELETE TOTALLY\n",
      "TO DELETE TOTALLY\n",
      "TO DELETE TOTALLY\n",
      "TO DELETE TOTALLY\n",
      "DELETE THE BEGIN OF EXON\n",
      "***\n",
      "DELETE THE END OF EXON\n",
      "DELETE THE BEGIN OF EXON\n",
      "***\n",
      "DELETE INSIDE THE EXON\n",
      "***\n",
      "DELETE INSIDE THE EXON\n",
      "***\n",
      "DELETE THE END OF EXON\n",
      "DELETE THE BEGIN OF EXON\n",
      "***\n",
      "DELETE THE END OF EXON\n",
      "TO DELETE TOTALLY\n",
      "***\n",
      "DELETE THE END OF EXON\n",
      "DELETE THE BEGIN OF EXON\n",
      "***\n",
      "TO DELETE TOTALLY\n",
      "DELETE THE BEGIN OF EXON\n",
      "***\n"
     ]
    }
   ],
   "source": [
    "# LOCALISATION EXACT DES MISMATCH\n",
    "\n",
    "from Bio.Data import CodonTable\n",
    "from Bio.Seq import translate\n",
    "\n",
    "for index, row in mismatch.iloc[:10,:].iterrows():\n",
    "    transcript_map = prim_exon_introns.loc[(prim_exon_introns[\"mismatch_ID\"] == int(row[0])) & (prim_exon_introns[\"type\"] == \"Exon\")]\n",
    "    id_exon_start = row[7]-1\n",
    "    id_exon_stop = row[8]\n",
    "    mismatch_peptid = row[11]\n",
    "    \n",
    "    # LOCALISATION EXACT DES MISMATCH\n",
    "    mismatch_exon_table = transcript_map.iloc[id_exon_start:id_exon_stop]\n",
    "    mismatch_exon = transcript_map.iloc[id_exon_start:id_exon_stop, 4]\n",
    "    exon_seq = mismatch_exon.str.cat(sep=\"\")\n",
    "    for i in range(3):\n",
    "        peptid = translate(exon_seq[i:])\n",
    "        frame = i\n",
    "        match_find = peptid.find(row[11])\n",
    "        if match_find != -1:\n",
    "            break\n",
    "    start_nuc_mismatch = match_find*3+frame\n",
    "    stop_nuc_mismatch = match_find*3+len(mismatch_peptid*3)+frame\n",
    "    # CREATION DES COORDONEE NUCLEOTIDES DES EXONS A MISMATCH\n",
    "    taille_exon = []\n",
    "    for index2, row2 in mismatch_exon_table.iloc[:,:].iterrows():\n",
    "        taille_exon.append(len(row2[4]))\n",
    "    taille_exon_fin = [sum(taille_exon[:x+1]) for x in range(0,len(taille_exon))]\n",
    "    taille_exon_debut = taille_exon_fin.copy()\n",
    "    taille_exon_debut.pop()\n",
    "    taille_exon_debut[:]=[i+1 for i in taille_exon_debut]\n",
    "    taille_exon_debut.insert(0, 0)\n",
    "    mismatch_exon_table[\"pos_start\"] = taille_exon_debut\n",
    "    mismatch_exon_table[\"pos_stop\"] = taille_exon_fin\n",
    "    \n",
    "    # SUPPRESSION DES PARTIE D'EXON DES TRANSCRIPT\n",
    "    for index2, row2 in mismatch_exon_table.iloc[:,:].iterrows():\n",
    "        if row2[5] >= start_nuc_mismatch and row2[6] <= stop_nuc_mismatch:\n",
    "            print(\"TO DELETE TOTALLY\")\n",
    "\n",
    "        if row2[5] < start_nuc_mismatch and row2[6] < stop_nuc_mismatch:\n",
    "            print(\"DELETE THE END OF EXON\")\n",
    "\n",
    "        if row2[5] > start_nuc_mismatch and row2[6] > stop_nuc_mismatch:\n",
    "            print(\"DELETE THE BEGIN OF EXON\")\n",
    "\n",
    "        if row2[5] <= start_nuc_mismatch and row2[6] >= stop_nuc_mismatch:\n",
    "            print(\"DELETE INSIDE THE EXON\")\n",
    "    \n",
    "    print(\"***\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 172,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DELETE THE END OF EXON\n",
      "DELETE THE BEGIN OF EXON\n"
     ]
    }
   ],
   "source": [
    "# SUPPRESSION DES PARTIE D'EXON DES TRANSCRIPT\n",
    "for index2, row2 in mismatch_exon_table.iloc[:,:].iterrows():\n",
    "    if row2[5] >= start_nuc_mismatch and row2[6] <= stop_nuc_mismatch:\n",
    "        print(\"TO DELETE TOTALLY\")\n",
    "    \n",
    "    if row2[5] <= start_nuc_mismatch and row2[6] <= stop_nuc_mismatch:\n",
    "        print(\"DELETE THE END OF EXON\")\n",
    "        \n",
    "    if row2[5] >= start_nuc_mismatch and row2[6] >= stop_nuc_mismatch:\n",
    "        print(\"DELETE THE BEGIN OF EXON\")\n",
    "\n",
    "    if row2[5] <= start_nuc_mismatch and row2[6] >= stop_nuc_mismatch:\n",
    "        print(\"DELETE INSIDE THE EXON\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.7.6 64-bit ('stage-env': conda)",
   "language": "python",
   "name": "python37664bitstageenvconda39052ae7dd75404f9bdb3226673ac6d7"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
