{
 "nbformat": 4,
 "nbformat_minor": 2,
 "metadata": {
  "language_info": {
   "name": "python",
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "version": "3.7.6-final"
  },
  "orig_nbformat": 2,
  "file_extension": ".py",
  "mimetype": "text/x-python",
  "name": "python",
  "npconvert_exporter": "python",
  "pygments_lexer": "ipython3",
  "version": 3,
  "kernelspec": {
   "name": "python37664bitstageenvcondaf79505b32b884913903330abaa217b9c",
   "display_name": "Python 3.7.6 64-bit ('stage-env': conda)"
  }
 },
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sqlite3\n",
    "import pandas as pd\n",
    "import os\n",
    "import sys\n",
    "from Bio.SubsMat import MatrixInfo as matlist\n",
    "from Bio import pairwise2\n",
    "from Bio.pairwise2 import format_alignment\n",
    "conn = sqlite3.connect('../../mismatch_db.db')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mismatch = pd.read_sql_query(\"SELECT * FROM mismatch\", conn)\n",
    "mismatch = mismatch.astype({\"exon_start_prim\": \"Int64\", \"exon_stop_prim\":\"Int64\", \"exon_start_hum\": \"Int64\", \"exon_stop_hum\":\"Int64\"})\n",
    "\n",
    "n_in_genom = pd.read_sql_query(\"\"\"\n",
    "SELECT mismatch.*, protein.prot_ID, transcript.sequence_genomic\n",
    "FROM mismatch\n",
    "JOIN protein ON mismatch.prot_prim = protein.prot_ID\n",
    "JOIN transcript ON protein.transcript_ID = transcript.transcript_ID\n",
    "\"\"\", conn)\n",
    "n_in_genom = n_in_genom.astype({\"exon_start_prim\": \"Int64\", \"exon_stop_prim\":\"Int64\", \"exon_start_hum\": \"Int64\", \"exon_stop_hum\":\"Int64\"})\n",
    "\n",
    "prim_introns = pd.read_sql_query(\"\"\"\n",
    "SELECT mismatch_ID, mismatch.prot_prim, exon_intron_map.'type', exon_intron_map.number_elem, exon_intron_map.seq\n",
    "FROM mismatch\n",
    "JOIN protein ON mismatch.prot_prim = protein.prot_ID\n",
    "JOIN transcript ON protein.transcript_ID = transcript.transcript_ID\n",
    "JOIN exon_intron_map ON transcript.transcript_ID = exon_intron_map.transcript_ID  WHERE exon_intron_map.'type' = 'Intron'\n",
    "\"\"\", conn)\n",
    "\n",
    "human_introns = pd.read_sql_query(\"\"\"\n",
    "SELECT mismatch_ID, mismatch.prot_hum, exon_intron_map.'type', exon_intron_map.number_elem, exon_intron_map.seq\n",
    "FROM mismatch\n",
    "JOIN protein ON mismatch.prot_hum = protein.prot_ID\n",
    "JOIN transcript ON protein.transcript_ID = transcript.transcript_ID\n",
    "JOIN exon_intron_map ON transcript.transcript_ID = exon_intron_map.transcript_ID  WHERE exon_intron_map.'type' = 'Intron'\n",
    "\"\"\", conn)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Flagging des mismatch conservés V\n",
    "counter = 0\n",
    "total = 0\n",
    "for index, row in error_list.iloc[:,:].iterrows():\n",
    "    \n",
    "    total+=1\n",
    "    conserved = 0\n",
    "\n",
    "    mySeq = fasta2List(\"../../data/raw/uniprot-sequence/\"+row[0][19:-6])\n",
    "    protein_Mismatch = [val for key, val in mySeq.items() if row[2] in key]\n",
    "\n",
    "    mismtaching_seq = protein_Mismatch[0][row[3]:row[4]]\n",
    "    for i, j in mySeq.items():\n",
    "        myAlign = pairwise2.align.localms(j, mismtaching_seq, 2, 0, -.5, -.1, one_alignment_only=True, score_only=True)\n",
    "        #print(format_alignment(*myAlign[0]))\n",
    "        if myAlign/len(mismtaching_seq) > 1.8:\n",
    "            conserved +=1\n",
    "\n",
    "    if conserved >= 4:\n",
    "        #print(str(index)+'\\t'.join([str(x) for x in row]))\n",
    "        counter +=1\n",
    "print(counter)\n",
    "print(total)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Flagging des erreurs d'alignement V\n",
    "counter = 0\n",
    "total = 0\n",
    "for index, row in error_list.iloc[:,:].iterrows():\n",
    "\n",
    "    mySeq = fasta2List(\"../../data/raw/uniprot-sequence/\"+row[0][19:-6])\n",
    "    protein_Mismatch = [val for key, val in mySeq.items() if row[2] in key]\n",
    "    human_prot_Mismatch = [val for key, val in mySeq.items() if row[0][20:-16] in key]\n",
    "    if human_prot_Mismatch == []:\n",
    "        pass\n",
    "    else:\n",
    "        total +=1\n",
    "        mismtaching_seq = protein_Mismatch[0]\n",
    "        human_seq = human_prot_Mismatch[0][row[5]:row[6]]\n",
    "\n",
    "        myAlign = pairwise2.align.localms(mismtaching_seq, human_seq, 2, -1, -.5, -.1, one_alignment_only=True, score_only=True)\n",
    "        if myAlign/len(human_seq) > 1.8:\n",
    "            #print(str(index)+'\\t'.join([str(x) for x in row]))\n",
    "            counter += 1\n",
    "\n",
    "print(counter)\n",
    "print(total)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Flagging des seq à repeats proteiq V\n",
    "counter = 0\n",
    "total = 0\n",
    "for index, row in error_list.iloc[:,:].iterrows():\n",
    "\n",
    "    mySeq = fasta2List(\"../../data/raw/uniprot-sequence/\"+row[0][19:-6])\n",
    "    protein_Mismatch = [val for key, val in mySeq.items() if row[2] in key]\n",
    "    human_prot_Mismatch = [val for key, val in mySeq.items() if row[0][20:-16] in key]\n",
    "    if human_prot_Mismatch == []:\n",
    "        pass\n",
    "    else:\n",
    "        total+=1\n",
    "        mismtaching_seq = protein_Mismatch[0]\n",
    "        human_seq = human_prot_Mismatch[0][row[5]:row[6]]\n",
    "\n",
    "        begin_seq = human_prot_Mismatch[0][:row[5]+1]\n",
    "        end_seq = human_prot_Mismatch[0][row[6]+2:]\n",
    "\n",
    "        human_prot_Mismatch = begin_seq + end_seq\n",
    "        myAlign = pairwise2.align.localms(human_prot_Mismatch, human_seq, 2, -1, -.5, -.1, one_alignment_only=True, score_only=True)\n",
    "        if myAlign/len(human_seq) > 1.8:\n",
    "            counter += 1\n",
    "\n",
    "print(counter)\n",
    "print(total)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Flagging des seq à N V\n",
    "\n",
    "\n",
    "counter = 0\n",
    "total = 0\n",
    "for index, row in error_list.iloc[:,:].iterrows():\n",
    "    my_CDS = []\n",
    "    subset_exon_pos = myExonicMap.loc[(myExonicMap[0]==row[0]) & (myExonicMap[2]==row[2]) & \\\n",
    "                      (myExonicMap[3]== str(row[3])) & (myExonicMap[4]== str(row[4]))]\n",
    "    subset_exon = exon_file.loc[exon_file[0]== row[2]]\n",
    "    subset_intron = intron_file.loc[intron_file[0]== row[2]]\n",
    "    try:\n",
    "        for n in range(int(subset_exon_pos[5]), int(subset_exon_pos[6])+1):\n",
    "            my_exon = subset_exon.loc[subset_exon[3] == n]\n",
    "            my_intron = subset_intron.loc[subset_intron[3] == n]\n",
    "            try:\n",
    "                my_CDS.append(my_exon.iloc[0,6])\n",
    "                my_CDS.append(my_intron.iloc[0,6])\n",
    "            except:\n",
    "                pass\n",
    "        \n",
    "        if \"N\" in ''.join(my_CDS):\n",
    "            counter +=1\n",
    "        total +=1\n",
    "    except:\n",
    "        pass\n",
    "\n",
    "print(counter)\n",
    "print(total)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Flagging des 1 exon human vs multiple primate V\n",
    "primate_exon_pos = pd.read_csv(\"../../data/mismatch-analysis2/mismatch_exon_pos.tab\", sep=\"\\t\")\n",
    "human_exon_pos = pd.read_csv(\"../../data/mismatch-flagging/human_mismatch_exon_pos.tab\", sep=\"\\t\")\n",
    "\n",
    "counter = 0\n",
    "total = 0\n",
    "for index, row in human_exon_pos.iloc[:, :].iterrows():\n",
    "    subset = primate_exon_pos.loc[primate_exon_pos[\"Alignement\"] == row[0]]\n",
    "    if subset.empty or row[7] == \"ERROR\":\n",
    "        continue\n",
    "    subset = subset.loc[subset[\"UniprotID\"] == row[2]]\n",
    "    subset = subset.loc[subset[\"PosStartError\"] == row[3]]\n",
    "    subset = subset.loc[subset[\"PosStopError\"] == row[4]]\n",
    "    try:\n",
    "        row_primate = subset.iloc[0,:].to_list()\n",
    "    except:\n",
    "        continue\n",
    "    total+=1\n",
    "    if (int(row[7])-int(row[8]) == 0) and (row_primate[6]-row_primate[5] >= 2):\n",
    "        #print(str(index)+'\\t'.join([str(x) for x in row]))\n",
    "        counter += 1\n",
    "\n",
    "print(counter)\n",
    "print(total)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Stat sur les sites introniques Humains\n",
    "\n",
    "canonical = 0\n",
    "non_canonical = 0\n",
    "human_intron_seq = pd.read_csv(\"../../data/mismatch-flagging/human_mismatch_intron_seq.tab\", sep=\"\\t\", header=None)\n",
    "for index, row in human_intron_seq.iloc[:, :].iterrows():\n",
    "    if row[6][:2] == \"GT\" and row[6][-2:] == \"AG\":\n",
    "        canonical += 1\n",
    "    #elif row[6][:2] == \"GC\" and row[6][-2:] == \"AG\":\n",
    "    #    canonical += 1\n",
    "    #elif row[6][:2] == \"AT\" and row[6][-2:] == \"AC\":\n",
    "    #    canonical += 1\n",
    "    else:\n",
    "        non_canonical += 1\n",
    "print(\"Canonical (GT/AG, GC/AG, AT/AC) \" + str(canonical/(canonical+non_canonical)*100))\n",
    "print(\"Non canonical \" + str(non_canonical/(canonical+non_canonical)*100))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Flagging des mismatch ayant un site humain de splicing non canonique V\n",
    "\n",
    "counter = 0\n",
    "total = 0\n",
    "for index, row in human_exon_pos.iloc[:, :].iterrows():\n",
    "    subset_intron = human_intron_map.loc[human_intron_map[0] == row[0][20:-15]]\n",
    "    if subset_intron.empty or row[7] == \"ERROR\":\n",
    "        continue\n",
    "    total += 1\n",
    "    for i in range(int(row[7]),int(row[8])+2):\n",
    "        intron = subset_intron.loc[subset_intron[3] == (i-1)]\n",
    "        if intron.empty:\n",
    "            continue\n",
    "        intron = intron.iloc[0,:].to_list()\n",
    "        #if  (intron[6][:2] == \"GT\" and intron[6][-2:] == \"AG\") \\\n",
    "        #    or (intron[6][:2] == \"GC\" and intron[6][-2:] == \"AG\") \\\n",
    "        #    or (intron[6][:2] == \"AT\" and intron[6][-2:] == \"AC\") and (len(intron[6]) > 30):\n",
    "        if  (intron[6][:2] == \"GT\" and intron[6][-2:] == \"AG\") and (len(intron[6]) > 30):\n",
    "            pass\n",
    "        else:\n",
    "            counter +=1\n",
    "            #print(str(index)+'\\t'.join([str(x) for x in row]))\n",
    "            break        \n",
    "\n",
    "print(counter)\n",
    "print(total)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Flagging mismatch with too small introns V\n",
    "counter = 0\n",
    "total = 0\n",
    "\n",
    "for index, row in primate_exon_pos.iloc[:, :].iterrows():\n",
    "    subset_intron = primate_intron_seq.loc[primate_intron_seq[0] == row[2]]\n",
    "    if subset_intron.empty or row[5] == \"ERROR\":\n",
    "        error+=1\n",
    "        continue\n",
    "    total += 1\n",
    "    for i in range(int(row[5]),int(row[6])):\n",
    "        intron = subset_intron.loc[subset_intron[3] == i]\n",
    "        intron = intron.iloc[0,:].to_list()\n",
    "        if len(intron[6]) <= 29:\n",
    "            counter +=1\n",
    "            #print(str(index)+'\\t'.join([str(x) for x in row]))\n",
    "            break\n",
    "\n",
    "print(counter)\n",
    "print(total)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Flagging isoform\n",
    "\n",
    "import requests\n",
    "import pandas as pd\n",
    "import sys\n",
    "\n",
    "URL = \"https://www.uniprot.org/uploadlists/\"\n",
    "\n",
    "params = {\n",
    "    \"from\": \"ACC+ID\",\n",
    "    \"to\": \"ENSEMBL_ID\",\n",
    "    \"format\": \"tab\",\n",
    "    \"query\": \"Q6H8Q1\"\n",
    "}\n",
    "\n",
    "r = requests.post(URL, data=params)\n",
    "print(r.text)\n",
    "\n",
    "params = {\n",
    "    \"from\": \"ENSEMBL_TRS_ID\",\n",
    "    \"to\": \"ACC\",\n",
    "    \"format\": \"tab\",\n",
    "    \"query\": 'ENST00000493231 ENST00000465414 ENST00000476906'\n",
    "}\n",
    "\n",
    "r = requests.post(URL, data=params)\n",
    "print(r.text)\n",
    "\"\"\""
   ]
  }
 ]
}