{
 "nbformat": 4,
 "nbformat_minor": 2,
 "metadata": {
  "language_info": {
   "name": "python",
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "version": "3.7.6-final"
  },
  "orig_nbformat": 2,
  "file_extension": ".py",
  "mimetype": "text/x-python",
  "name": "python",
  "npconvert_exporter": "python",
  "pygments_lexer": "ipython3",
  "version": 3,
  "kernelspec": {
   "name": "python37664bitstageenvcondaf79505b32b884913903330abaa217b9c",
   "display_name": "Python 3.7.6 64-bit ('stage-env': conda)"
  }
 },
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sqlite3\n",
    "import pandas as pd\n",
    "import os\n",
    "import sys\n",
    "from Bio.SubsMat import MatrixInfo as matlist\n",
    "from Bio import pairwise2\n",
    "from Bio.pairwise2 import format_alignment\n",
    "import math\n",
    "import numpy as np\n",
    "def fasta2List(pathFasta):\n",
    "    f = open(pathFasta, \"r\")\n",
    "    title = []\n",
    "    seq = []\n",
    "    seq_temp = []\n",
    "    for line in f:\n",
    "        if line[0] == \">\":\n",
    "            seq.append(''.join(seq_temp).replace(\"\\n\", \"\"))\n",
    "            title.append(line.replace(\"\\n\", \"\"))\n",
    "            seq_temp = []\n",
    "        else:\n",
    "            seq_temp.append(line)\n",
    "    seq.append(''.join(seq_temp).replace(\"\\n\", \"\"))\n",
    "    seq.pop(0)\n",
    "    dictionary = dict(zip(title, seq))\n",
    "    return dictionary\n",
    "\n",
    "conn = sqlite3.connect('../../mismatch_db.db')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mismatch = pd.read_sql_query(\"SELECT * FROM mismatch\", conn)\n",
    "mismatch = mismatch.astype({\"exon_start_prim\": \"Int64\", \"exon_stop_prim\":\"Int64\", \"exon_start_hum\": \"Int64\", \"exon_stop_hum\":\"Int64\"})\n",
    "\n",
    "prot_seq_hum = pd.read_sql_query(\"\"\"\n",
    "SELECT mismatch.mismatch_ID, mismatch.prot_hum, protein.sequence\n",
    "FROM mismatch\n",
    "JOIN protein ON mismatch.prot_hum = protein.prot_ID\"\"\", conn)\n",
    "\n",
    "prot_seq_prim = pd.read_sql_query(\"\"\"\n",
    "SELECT mismatch.mismatch_ID, mismatch.prot_prim, protein.sequence\n",
    "FROM mismatch\n",
    "JOIN protein ON mismatch.prot_prim = protein.prot_ID\"\"\", conn)\n",
    "\n",
    "prim_exon_introns = pd.read_sql_query(\"\"\"\n",
    "SELECT mismatch_ID, mismatch.prot_prim, exon_intron_map.'type', exon_intron_map.number_elem, exon_intron_map.seq\n",
    "FROM mismatch\n",
    "JOIN protein ON mismatch.prot_prim = protein.prot_ID\n",
    "JOIN transcript ON protein.transcript_ID = transcript.transcript_ID\n",
    "JOIN exon_intron_map ON transcript.transcript_ID = exon_intron_map.transcript_ID\n",
    "\"\"\", conn)\n",
    "\n",
    "human_exon_introns = pd.read_sql_query(\"\"\"\n",
    "SELECT mismatch_ID, mismatch.prot_hum, exon_intron_map.'type', exon_intron_map.number_elem, exon_intron_map.seq\n",
    "FROM mismatch\n",
    "JOIN protein ON mismatch.prot_hum = protein.prot_ID\n",
    "JOIN transcript ON protein.transcript_ID = transcript.transcript_ID\n",
    "JOIN exon_intron_map ON transcript.transcript_ID = exon_intron_map.transcript_ID\n",
    "\"\"\", conn)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Flagging des mismatch conservés V\n",
    "counter = 0\n",
    "total = 0\n",
    "index_conserved = []\n",
    "for index, row in mismatch.iloc[:,:].iterrows():\n",
    "    total+=1\n",
    "    conserved = 0\n",
    "\n",
    "    mySeq = fasta2List(\"../../data/raw/uniprot-sequence/\"+row[1]+\".id.fasta\")\n",
    "    mismtaching_seq = row[11]\n",
    "\n",
    "    for i, j in mySeq.items():\n",
    "        myAlign = pairwise2.align.localms(j, mismtaching_seq, 2, 0, -.5, -.1, one_alignment_only=True, score_only=True)\n",
    "        #print(format_alignment(*myAlign[0]))\n",
    "        if myAlign/len(mismtaching_seq) > 1.8:\n",
    "            conserved +=1\n",
    "\n",
    "    if conserved >= 4:\n",
    "        index_conserved.append(index)\n",
    "        #print(str(index)+'\\t'.join([str(x) for x in row]))\n",
    "        counter +=1\n",
    "print(counter)\n",
    "print(total)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Flagging des erreurs d'alignement V\n",
    "counter = 0\n",
    "total = 0\n",
    "index_align_error = []\n",
    "for index, row in mismatch.iloc[:,:].iterrows():\n",
    "    try:\n",
    "        human_seq = prot_seq_hum.loc[prot_seq_hum[\"prot_hum\"]== row[1]].iloc[0,2]\n",
    "    except:\n",
    "        continue\n",
    "    total +=1\n",
    "    mismtaching_seq = row[11]\n",
    "    myAlign = pairwise2.align.localms(human_seq, mismtaching_seq, 2, 0, -.5, -.1, one_alignment_only=True, score_only=True)\n",
    "    if myAlign/len(mismtaching_seq) > 1.8:\n",
    "        index_align_error.append(index)\n",
    "        #print(str(index)+'\\t'.join([str(x) for x in row]))\n",
    "        counter += 1\n",
    "\n",
    "print(counter)\n",
    "print(total)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Flagging des seq à repeats proteiq V\n",
    "counter = 0\n",
    "total = 0\n",
    "index_repeats = []\n",
    "for index, row in mismatch.iloc[:,:].iterrows():\n",
    "    try:\n",
    "        human_prot_Mismatch = prot_seq_hum.loc[prot_seq_hum[\"prot_hum\"]== row[1]].iloc[0,2]\n",
    "    except:\n",
    "        continue\n",
    "\n",
    "    total+=1\n",
    "    human_seq = row[12]\n",
    "    begin_seq = human_prot_Mismatch[:row[5]+1]\n",
    "    end_seq = human_prot_Mismatch[row[6]+2:]\n",
    "    human_prot_Mismatch = begin_seq + end_seq\n",
    "    myAlign = pairwise2.align.localms(human_prot_Mismatch, human_seq, 2, 0, -.5, -.1, one_alignment_only=True, score_only=True)\n",
    "    if myAlign/len(human_seq) > 1.8:\n",
    "        index_repeats.append(index)\n",
    "        counter += 1\n",
    "\n",
    "print(counter)\n",
    "print(total)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Flagging des seq à N \n",
    "\n",
    "counter = 0\n",
    "total = 0\n",
    "index_genom_n = []\n",
    "for index, row in mismatch.iloc[:,:].iterrows():\n",
    "    my_CDS = []\n",
    "    subset_exon_intron = prim_exon_introns.loc[prim_exon_introns[\"mismatch_ID\"] == index]\n",
    "    if isinstance(row[7], int) == False or isinstance(row[8], int) == False:\n",
    "        continue\n",
    "    try:\n",
    "        for n in range(row[7], row[8]+1):\n",
    "            my_exon = subset_exon_intron.loc[(subset_exon_intron[\"number_elem\"] == n) & (subset_exon_intron[\"type\"] == \"Exon\")]\n",
    "            my_intron = subset_exon_intron.loc[(subset_exon_intron[\"number_elem\"] == n) & (subset_exon_intron[\"type\"] == \"Intron\")]\n",
    "            try:\n",
    "                my_CDS.append(my_exon.iloc[0,4])\n",
    "                my_CDS.append(my_intron.iloc[0,4])\n",
    "            except:\n",
    "                pass\n",
    "        if \"N\" in ''.join(my_CDS):\n",
    "            index_genom_n.append(index)\n",
    "            counter +=1\n",
    "        total +=1\n",
    "    except:\n",
    "        pass\n",
    "\n",
    "print(counter)\n",
    "print(total)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Flagging des 1 exon human vs multiple primate V\n",
    "counter = 0\n",
    "total = 0\n",
    "index_multiple_exon = []\n",
    "for index, row in mismatch.iloc[:, :].iterrows():\n",
    "    if isinstance(row[7], int) == False or isinstance(row[9], int) == False:\n",
    "        continue\n",
    "    total +=1\n",
    "    if (int(row[9])-int(row[10]) == 0) and (row[8]-row[7] >= 2):\n",
    "        #print(str(index)+'\\t'.join([str(x) for x in row]))\n",
    "        index_multiple_exon.append(index)\n",
    "        counter += 1\n",
    "\n",
    "print(counter)\n",
    "print(total)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Stat sur les sites introniques Humains\n",
    "\n",
    "canonical = 0\n",
    "non_canonical = 0\n",
    "human_intron_seq = human_exon_introns.loc[human_exon_introns[\"type\"]==\"Intron\"]\n",
    "for index, row in human_intron_seq.iloc[:, [0,4]].iterrows():\n",
    "    if row[1][:2] == \"GT\" and row[1][-2:] == \"AG\":\n",
    "        canonical += 1\n",
    "    #elif row[6][:2] == \"GC\" and row[6][-2:] == \"AG\":\n",
    "    #    canonical += 1\n",
    "    #elif row[6][:2] == \"AT\" and row[6][-2:] == \"AC\":\n",
    "    #    canonical += 1\n",
    "    else:\n",
    "        non_canonical += 1\n",
    "print(\"Canonical (GT/AG) \" + str(canonical/(canonical+non_canonical)*100))\n",
    "print(\"Non canonical \" + str(non_canonical/(canonical+non_canonical)*100))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mismatch"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "# Flagging des mismatch ayant un site humain de splicing non canonique V\n",
    "counter = 0\n",
    "total = 0\n",
    "index_non_canon = []\n",
    "for index, row in mismatch.iloc[:, :].iterrows():\n",
    "    subset_intron = human_exon_introns.loc[(human_exon_introns['mismatch_ID'] == row[0]) & (human_exon_introns['type'] == \"Intron\")]\n",
    "    if subset_intron.empty or row[9] == \"ERROR\":\n",
    "        continue\n",
    "    total += 1\n",
    "    for i in range(int(row[9]),int(row[10])+2):\n",
    "        intron = subset_intron.loc[subset_intron[\"number_elem\"] == (i-1)]\n",
    "        if intron.empty:\n",
    "            continue\n",
    "        intron = intron.iloc[0,:].to_list()\n",
    "        #if  (intron[6][:2] == \"GT\" and intron[6][-2:] == \"AG\") \\\n",
    "        #    or (intron[6][:2] == \"GC\" and intron[6][-2:] == \"AG\") \\\n",
    "        #    or (intron[6][:2] == \"AT\" and intron[6][-2:] == \"AC\") and (len(intron[6]) > 30):\n",
    "        if  (intron[4][:2] == \"GT\" and intron[4][-2:] == \"AG\") and (len(intron[4]) > 30):\n",
    "            pass\n",
    "        else:\n",
    "            index_non_canon.append(index)\n",
    "            counter +=1\n",
    "            #print(str(index)+'\\t'.join([str(x) for x in row]))\n",
    "            break        \n",
    "\n",
    "print(counter)\n",
    "print(total)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Flagging mismatch with too small introns V\n",
    "counter = 0\n",
    "total = 0\n",
    "index_intron_small = []\n",
    "for index, row in mismatch.iloc[:, :].iterrows():\n",
    "    subset_intron = prim_exon_introns.loc[(prim_exon_introns['mismatch_ID'] == row[0]) & (prim_exon_introns['type'] == \"Intron\")]\n",
    "    if subset_intron.empty or row[7] == \"ERROR\":\n",
    "        continue\n",
    "    total += 1\n",
    "    for i in range(int(row[7]),int(row[8])):\n",
    "        intron = subset_intron.loc[subset_intron[\"number_elem\"] == i]\n",
    "        intron = intron.iloc[0,:].to_list()\n",
    "        if len(intron[4]) <= 29:\n",
    "            index_intron_small.append(index)\n",
    "            counter +=1\n",
    "            #print(str(index)+'\\t'.join([str(x) for x in row]))\n",
    "            break\n",
    "\n",
    "print(counter)\n",
    "print(total)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "column_names = [\"mismatch_ID\", \"conserved\", \"one_hum_multiple_prim\", \"non_canonical_hum_spl\", \"N_in_genomic\", \"small_introns\", \"repeats_prot\", \"alignement_error\", \"human_isoform_exist\"]\n",
    "df = pd.DataFrame(columns = column_names)\n",
    "df[\"mismatch_ID\"] = mismatch[\"mismatch_ID\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df['conserved'] = np.where(df[\"mismatch_ID\"].isin(index_conserved), 1, 0)\n",
    "df['one_hum_multiple_prim'] = np.where(df[\"mismatch_ID\"].isin(index_multiple_exon), 1, 0)\n",
    "df['non_canonical_hum_spl'] = np.where(df[\"mismatch_ID\"].isin(index_non_canon), 1, 0)\n",
    "df['N_in_genomic'] = np.where(df[\"mismatch_ID\"].isin(index_genom_n), 1, 0)\n",
    "df['small_introns'] = np.where(df[\"mismatch_ID\"].isin(index_intron_small), 1, 0)\n",
    "df['repeats_prot'] = np.where(df[\"mismatch_ID\"].isin(index_repeats), 1, 0)\n",
    "df['alignement_error'] = np.where(df[\"mismatch_ID\"].isin(index_align_error), 1, 0)\n",
    "#df['human_isoform_exist'] = np.where(df[\"mismatch_ID\"].isin(index_conserved), 1, 0)\n",
    "#df.to_sql(con=conn, name='mismatch_flag', index=False, if_exists=\"append\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tata = pd.read_sql_query(\n",
    "\"\"\"SELECT * FROM mismatch_flag \n",
    "WHERE conserved = 0 \n",
    "AND one_hum_multiple_prim = 0 \n",
    "AND non_canonical_hum_spl = 0\n",
    "AND N_in_genomic = 0\n",
    "AND small_introns = 0\n",
    "AND repeats_prot = 0\n",
    "AND alignement_error = 0\n",
    "\"\"\", conn)\n",
    "tata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Flagging isoform\n",
    "# Flagging isoform\n",
    "\n",
    "import requests\n",
    "import pandas as pd\n",
    "import sys\n",
    "import json\n",
    "URL = \"https://www.uniprot.org/uploadlists/\"\n",
    "\n",
    "# Prot to gene\n",
    "params = {\n",
    "    \"from\": \"ACC+ID\",\n",
    "    \"to\": \"ENSEMBL_ID\",\n",
    "    \"format\": \"tab\",\n",
    "    \"query\": \"Q6H8Q1\"\n",
    "}\n",
    "\n",
    "r = requests.post(URL, data=params)\n",
    "print(r.text)\n",
    "\n",
    "# Gene to Transcript\n",
    "url = \"https://rest.ensembl.org/lookup/id/\"\n",
    "headers = {\"Content-Type\": \"application/json\",\n",
    "            \"Accept\": \"application/json\"}\n",
    "\n",
    "params = {\"ids\": [\"ENSG00000163995\"], \"expand\": \"1\"}\n",
    "r = requests.post(URL, headers=headers, data=json.dumps(params))\n",
    "\n",
    "# Transcript to prot\n",
    "params = {\n",
    "    \"from\": \"ENSEMBL_TRS_ID\",\n",
    "    \"to\": \"ACC\",\n",
    "    \"format\": \"tab\",\n",
    "    \"query\": 'ENST00000493231 ENST00000465414 ENST00000476906'\n",
    "}\n",
    "\n",
    "r = requests.post(URL, data=params)\n",
    "print(r.text)"
   ]
  }
 ]
}