{
 "nbformat": 4,
 "nbformat_minor": 2,
 "metadata": {
  "language_info": {
   "name": "python",
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "version": "3.7.6-final"
  },
  "orig_nbformat": 2,
  "file_extension": ".py",
  "mimetype": "text/x-python",
  "name": "python",
  "npconvert_exporter": "python",
  "pygments_lexer": "ipython3",
  "version": 3,
  "kernelspec": {
   "name": "python3",
   "display_name": "Python 3"
  }
 },
 "cells": [
  {
   "cell_type": "markdown",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Script permettant de tenter de corriger un mismatch en alignant le peptide human a la seq génomic et en cherchant le meilleur match\n",
    "Input:   \n",
    "    ID_file fichier de correspondance UniprotID - Ensembl Trasncript ID  \n",
    "    Error_file fichier contenant toute les erreurs de type mismatch (script julie)  \n",
    "    Exon_file fichier contenant les informations sur les positions des exons des transcripts  \n",
    "    my_Genomic fichier fasta contenant les séquences génomiques des transcripts  \n",
    "    my_cDNA fichier fasta contenant les séquences cDNA des transcripts  \n",
    "    my_CDS fichier fasta contenant les séquences CDS des transcripts "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Importer les lib et fonction\n",
    "import json\n",
    "import pandas as pd\n",
    "\n",
    "def fasta2List(pathFasta):\n",
    "    f = open(pathFasta, \"r\")\n",
    "    title = []\n",
    "    seq = []\n",
    "    seq_temp = []\n",
    "    for line in f:\n",
    "        if line[0] == \">\":\n",
    "            seq.append(''.join(seq_temp).replace(\"\\n\", \"\"))\n",
    "            title.append(line.replace(\"\\n\", \"\"))\n",
    "            seq_temp = []\n",
    "        else:\n",
    "            seq_temp.append(line)\n",
    "    seq.append(''.join(seq_temp).replace(\"\\n\", \"\"))\n",
    "    seq.pop(0)\n",
    "    dictionary = dict(zip(title, seq))\n",
    "    return dictionary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Importation de toutes les données utilisées\n",
    "ID_file = pd.read_csv(\"../raw/uniprot-exon-map/transcript_ensembl.tab\", sep = \"\\t\")\n",
    "Error_file = pd.read_csv(\"../raw/uniprot-exon-map/uniprot_new_errors_filt.txt\", sep=\" \", header=None)\n",
    "Exon_file = pd.read_csv(\"../raw/uniprot-exon-map/Exon_map.tab\", sep=\"\\t\", header=None)\n",
    "my_Genomic = fasta2List(\"../raw/uniprot-exon-map/genomics_new.fa\")\n",
    "my_cDNA = fasta2List(\"../raw/uniprot-exon-map/cdna_seq.fa\")\n",
    "my_CDS = fasta2List(\"../raw/uniprot-exon-map/cds_new.fa\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Pour chaque mismatch du fichier d'erreur: on cherche la séquence humaine en face du mismatch. On prend le séquence génomique de la protéine du primate\n",
    "# concerné et on la traduit dans les trois cadre de lecture. On cherche si la séquence humaine peut être trouvée dans une des séquence génomique traduite\n",
    "# Si oui on écrit dans un fichier  les informations du mismatch et la position génomique permetttant de le régler. \n",
    "from Bio.Seq import Seq\n",
    "from Bio.Alphabet import generic_dna\n",
    "from Bio.Align.Applications import MafftCommandline\n",
    "\n",
    "for index, row in Error_file.iloc[:1,:].iterrows():\n",
    "    fasta_name = row[0][20:-6]\n",
    "    prot_name = row[2]\n",
    "    error_start = row[3]\n",
    "    error_stop = row[4]\n",
    "    human_start = row[5]\n",
    "    human_stop = row[6]\n",
    "\n",
    "    Prot_list = fasta2List(\"../raw/uniprot-sequence/\"+fasta_name)\n",
    "    prot_HumanRef = [val for key, val in Prot_list.items() if row[0][20:-15] in key]\n",
    "    genomic_Seq = [val for key, val in my_Genomic.items() if prot_name in key]\n",
    "    \n",
    "    if prot_HumanRef == []:\n",
    "        pass\n",
    "    else:\n",
    "        peptide_Ref = (row[0][20:-15], prot_HumanRef[0][human_start:human_stop])\n",
    "        \n",
    "        # Translation & alignement part\n",
    "        querySeq = Seq(peptide_Ref[1])\n",
    "        prot_trial = []\n",
    "        for i in range(3):\n",
    "            f = open(\"../temp/\"+str(index)+\"_\"+peptide_Ref[0]+\"_\"+str(i)+\".fasta\", \"w\")\n",
    "            genomic_prot = Seq(genomic_Seq[0][i:], generic_dna)\n",
    "            f.write(\">\"+prot_name+\" genomic frame \"+str(i)+\"\\n\"+str(genomic_prot.translate())+\"\\n\")\n",
    "            f.write(\">\"+peptide_Ref[0]+\" Human Peptide query\"+\"\\n\"+str(peptide_Ref[1])+\"\\n\")\n",
    "            f.close()\n",
    "            \n",
    "            # Alignement MAFFT avec paramètre strictes sur l'ouverte et l'extension de gap\n",
    "            mafft_cline = MafftCommandline(input=\"../temp/\"+str(index)+\"_\"+peptide_Ref[0]+\"_\"+str(i)+\".fasta\", op=5.0, ep=10.0)\n",
    "            mafft_results = mafft_cline()\n",
    "            with open(\"../temp/\"+str(index)+\"_\"+peptide_Ref[0]+\"_\"+str(i)+\".fasta.mafft\", \"w\") as handle:\n",
    "                handle.write(mafft_results[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import AlignIO\n",
    "import os\n",
    "\n",
    "f = open(\"../raw/correction-pairwise/translation_match.tab\", \"w\")\n",
    "align_files = os.listdir(\"../raw/correction-pairwise/mafft/\")\n",
    "# Scan de la séquence aligné peptide humain pour trouver la position de début et de fin sur l'alignement\n",
    "f.write(\"Match\\tPrimate\\tHuman\\tStartPos\\tStopPos\\tSequence_Primate\\tSequence_Humaine\\n\")\n",
    "for align_file in align_files:\n",
    "    align = AlignIO.read(\"../raw/correction-pairwise/mafft/\"+align_file, \"fasta\")\n",
    "    start = 0\n",
    "    found_Start = False\n",
    "    for index, amino in enumerate(align[1, :].seq):\n",
    "        if amino.isalpha() and found_Start == False:\n",
    "            start = index\n",
    "            found_Start = True\n",
    "        if found_Start and amino.isalpha():\n",
    "            last_amino_index = index+1\n",
    "\n",
    "    peptide_genomique = align[0,start:last_amino_index].seq\n",
    "    peptide_humain_query = align[1,start:last_amino_index].seq\n",
    "\n",
    "    # Boucle qui scan l'identité entre la paire de peptide\n",
    "    identity_count = 0\n",
    "    gap_count = 0\n",
    "    for i in range(len(peptide_humain_query)):\n",
    "        if peptide_genomique[i] == peptide_humain_query[i]:\n",
    "            identity_count += 1\n",
    "        if (peptide_genomique[i] == \"-\") or (peptide_humain_query[i] == \"-\"):\n",
    "            gap_count += 1\n",
    "    identity_percentage = identity_count/len(peptide_humain_query)\n",
    "\n",
    "    if gap_count < 10 and identity_percentage > 0.8:\n",
    "        f.write(\"Match\\t\"+align[0].id+\"\\t\"+align[1].id+\"\\t\"+str(start*3)+\"\\t\"+str(last_amino_index*3)+\"\\t\"\n",
    "            +str(peptide_genomique)+\"\\t\"+str(peptide_humain_query)+\"\\n\")\n",
    "    else:\n",
    "        pass\n",
    "f.close()"
   ]
  }
 ]
}